### DISCUSSION TITLE
Outline view got to source not always correct

### DISCUSSION THREAD

**Discussion by sidprice** [microsoft/vscode-cpptools#7671] [Q&A]
Version: 1.56.2 (system setup)
Commit: 054a9295330880ed74ceaedda236253b4f39a335
Date: 2021-05-12T17:13:13.157Z
Electron: 12.0.4
Chrome: 89.0.4389.114
Node.js: 14.16.0
V8: 8.9.255.24-electron.0
OS: Windows_NT x64 10.0.19042

In some files if I double click on a function in the Outline view the editor window goes to the wrong place. How do I clear whatever cache is causing this, in general the go to source from the Outline view works well. Just a couple of files that seem "broken".

**Comment by sean-mcmanus** [comment]
There could be an error parsing the symbols in the file. If the file relies on a macro that breaks our parser, you may need to add the macro to a cpp.hint file (https://docs.microsoft.com/en-us/cpp/build/reference/hint-files?view=msvc-160).

Editing/saving the file should cause a refresh of the document symbols.

Are you able to provide a code sample of the issue?

**Reply by sidprice** [reply]
Hi, thanks for the quick response. I have attached a file that shows this problem, I hope it helps as a stand alone file.

The variable declarations starting around line #382 seem to trigger the failure.

[lp5036_driver.zip](https://github.com/microsoft/vscode-cpptools/files/6627013/lp5036_driver.zip)

**Reply by sean-mcmanus** [reply]
I'm not reproing any bug, i.e. document Outline appears to be working. Do you have 1.4.1 of the C/C++ extension installed?

Can you show a screenshot of what you're clicking and where it goes?

**Reply by sidprice** [reply]
Yes, that is the version of the extension.

attached is video capture.
[vscode_2021-06-09_162104.zip](https://github.com/microsoft/vscode-cpptools/files/6627182/vscode_2021-06-09_162104.zip)

**Reply by sean-mcmanus** [reply]
Do you know why the Outline only has methods and not macros, variables, etc.? This is what I'm seeing:

![image](https://user-images.githubusercontent.com/19859882/121438901-aeecb780-c939-11eb-9bdc-c9d60fd4305d.png)

Have you edited/saved the file? i.e. if you add a new new method does it update?

**Reply by sidprice** [reply]
ledBankNumbers is a variable, at the top of your screencap. I see a whole long list of them.

![vscode_variablesCapture](https://user-images.githubusercontent.com/12300040/121439363-0ba0a000-c943-11eb-8344-e307e48c877f.PNG)

I added a function but the issue did not change.

The file has been edited/saved/compiled many times. Still the issue remains.

**Reply by sidprice** [reply]
The added function and an added MACRO appear in the Outline and double-click takes you to them in the editor.

**Reply by sidprice** [reply]
BTW: This is the GCC-ARM compiler if that is significant?

**Reply by sean-mcmanus** [reply]
The Outline (document symbols) is independent of the compiler specifics, so gcc-arm shouldn't matter.

I don't know why certain methods would have the wrong line number associated with the symbols or what could cause it to not repro for me (I don't recall any similar bug reports like that). Maybe there database is somehow corrupt? You could try running the C/C++: Reset IntelliSense Database command to reset the entire database.

@Colengms Are you able to repro this?

**Reply by sean-mcmanus** [reply]
Oh -- what is your file encoding? UTF-8?

**Reply by sidprice** [reply]
Not sure on file encoding, how is that discovered please?

**Reply by sidprice** [reply]
Reset Intellisense DB and then a restart did the trick!

**Reply by sean-mcmanus** [reply]
It's in the bottom right status bar -- I re-watched your video and see it's UTF-8 so that should be okay.

Could you possibly have a different extension installed that could be providing those symbols? You can verify if our extension is being used because you should see no main() symbol with the code below (unless `#define foo` is added to a cpp.hint file):

```
#define foo

int main() foo
{
    return 0;
}
```

**Reply by sean-mcmanus** [reply]
Cool to hear that Reset fixed it -- we just use a sqlite database...not sure exactly how the database could be in a state where only some of the symbols in the file are invalid...hopefully that is a rare occurance.

**Reply by sidprice** [reply]
Thank you so much for being a great, responsive help. I love VSCode and this has been bugging me for quite a while now :)

**Reply by sean-mcmanus** [reply]
Yeah, no problem. I was thinking editing the file was supposed to update the database for all the symbols in the file, but I think the code got optimized to do a diff so only a subset of the document symbols get updated, which would explain why the Reset would be needed to refresh old symbol data. Let us know if you find a repro for getting the database into the bad state...seems like the file changed but the symbol incorrectly didn't get updated.

**Reply by sidprice** [reply]
Sure, I will try to note the trigger if I can :)
